### ISSUE TITLE
[rush] rush build --from does not build transitive dependencies

### ISSUE THREAD

**Issue by mikeharder** [microsoft/rushstack#1447]
## Overview
Consider a dependency graph where `B` depends on `A`, and `D` depends on both `B` and `C`:

```
A <- B <- D
     C <- D
```

Say I am making changes to `B`, and want to verify that `B` and all its downstream dependencies can build correctly.  I use the following command:

```
rush build --to B --from B
```

## Expected
Rush should build all projects downstream of `B`, including their transitive dependencies (which are required for them to build).  `D` is downstream of `B` so it must be built, and `C` is upstream of `D` so it must also be built.

## Actual
Rush does not build `C`.

## Workarounds
Without building the transitive dependencies of downstream projects, the only workaround I see is to build all projects in the monorepo.

## Repo
https://github.com/mikeharder/rush-to-from

**Comment by iclanton** [comment]
Looking into this. So the dependency graph for your repro repo is:

```
D --> C (not built)
  `--> B --> A
```

This illustrates a weirdness with the `--from` parameter, where Rush doesn't ensure the dependencies of projects it's building are built. I think we have two options for addressing this:

- Change the behavior of the `--from` parameter to also include all other dependencies of projects that are to be built. This will be a breaking change.
- Add the aforementioned behavior with an additional flag (something like `--ensure-dependencies-are-built`, but with a better name) to avoid making a breaking change.

I'm not sure which approach I prefer, or if there are any additional options.

@octogonz, @apostolisms - thoughts?

**Comment by mikeharder** [comment]
@iclanton: Thanks for taking a look.  I just noticed the title of this issue incorrectly said `--to` instead of `--from` (now corrected), but you are correct the issue is with `--from` not ensuring dependencies are built.

We would be fine with either option (changing `--from` or additional flag).  If one option could be implemented more quickly (say the additional flag since it's a non-breaking change), we would prefer that option.

**Comment by octogonz** [comment]
@nickpape-msft FYI

**Comment by octogonz** [comment]
@nickpape FYI

**Comment by octogonz** [comment]
This picture looks exactly correct for `rush build --to B --from B`:

```
D --> C (not built)
  `--> B --> A
```

- `--to B` means _"build B, plus the things that B depends on"_.  That set is `{ A, B }`.
- `--from B` means _"build B, plus all the things that depend on B"_  That set is `{ B, D }`.

When you specify both parameters, you get the union `{ A, B, D }`.

That union is meaningful and useful.  For example, suppose I did a clean clone of the repo, and I made some changes to B.  The  `--to B` ensures that I build `B` and all its prerequisites.  The `--from B` ensures that I run all the unit tests that could possibly by broken by this change.  In this scenario, building `C` is irrelevant and potentially would waste a lot of time.

**Comment by octogonz** [comment]
Ohhhh... wait a sec, I get it.  The problem is that `D` fails to build because `C` wasn't build.  Duh.  :-P

**Comment by octogonz** [comment]
Yeah, let's fix this.

**Comment by nthypes** [comment]
No updates?

**Comment by arikon** [comment]
What is the progress?

**Comment by octogonz** [comment]
This will be fixed by PR #2422

**Comment by mikeharder** [comment]
Closing issue since PR #2422 has been merged
